ecosystem/indexer-grpc/indexer-grpc-file-store/src/processor.rs

```diff
@@ -87,109 +84,174 @@ impl Processor {
-            tps_calculator.tick_now(process_size as u64);
-            let end_version = current_file_store_version + process_size as u64 - 1_u64;
-            let num_transactions = end_version - current_file_store_version + 1;
+                .is_err()
+            {
+                tracing::error!(
+                    batch_start_version = batch_start_version,
+                    "Failed to update file store metadata. Retrying."
+                );
+                std::thread::sleep(std::time::Duration::from_millis(500));
+                METADATA_UPLOAD_FAILURE_COUNT.inc();
+            }
+            info!(
+                end_version = batch_start_version,
+                duration_in_secs = start_time.elapsed().as_secs_f64(),
+                service_type = SERVICE_TYPE,
+                "{}",
+                IndexerGrpcStep::FilestoreUpdateMetadata.get_label()
+            );
+            LATEST_PROCESSED_VERSION
+                .with_label_values(&[
+                    SERVICE_TYPE,
+                    IndexerGrpcStep::FilestoreUpdateMetadata.get_step(),
+                    IndexerGrpcStep::FilestoreUpdateMetadata.get_label(),
+                ])
+                .set(batch_start_version as i64);
+            DURATION_IN_SECS
+                .with_label_values(&[
+                    SERVICE_TYPE,
+                    IndexerGrpcStep::FilestoreUpdateMetadata.get_step(),
+                    IndexerGrpcStep::FilestoreUpdateMetadata.get_label(),
+                ])
+                .set(start_time.elapsed().as_secs_f64());
+            let size = last_version - first_version + 1;
+            PROCESSED_VERSIONS_COUNT.inc_by(size);
+            tps_calculator.tick_now(size);
             // This decoding may be inefficient, but this is the file store so we don't have to be overly
             // concerned with efficiency.
             let end_version_timestamp = {
-                let encoded_transaction = last_transaction.0;
                 let decoded_transaction =
-                    base64::decode(encoded_transaction).expect("Failed to decode base64.");
+                    base64::decode(last_version_encoded).expect("Failed to decode base64.");
                 let transaction =
                     Transaction::decode(&*decoded_transaction).expect("Failed to decode protobuf.");
                 transaction.timestamp
             };
 
+            let duration = latest_time.elapsed().as_secs_f64();
             info!(
                 tps = (tps_calculator.avg() * 1000.0) as u64,
-                current_file_store_version = current_file_store_version,
-                "Upload transactions to file store."
+                start_version = first_version,
+                end_version = last_version,
+                duration_in_secs = duration,
+                service_type = SERVICE_TYPE,
+                "{}",
+                IndexerGrpcStep::FilestoreUploadTxns.get_label()
```

Comment:
This event should be moved to before updating metadata

Reply:
I guess this can be the e2e metric that includes the update metadata for now. Otherwise we need to add both steps just to see the overall duration.

Reply:
that's fine, then just rename the variable